ranking/src/ts_system/mod.rs

```diff
@@ -0,0 +1,303 @@
+type TSContestPlace = Vec<TSTeam>;
+type TSContest = Vec<TSContestPlace>;
+type TSRating = HashMap<TSPlayer, TSPlayerRating>;
+
+// TrueSkillStPB rating system
+pub struct TrueSkillSPBSystem {
+    // epsilon used for ties
+    eps: f64,
+    // performance sigma
+    beta: f64,
+    // epsilon used for convergence loop
+    convergence_eps: f64,
+    // defines sigma growth per second
+    sigma_growth: f64,
+    // history of ratings for all players
+    rating: TSRating,
+}
+
+impl Default for TrueSkillSPBSystem {
+    fn default() -> Self {
+        Self {
+            eps: 0.10,
+            beta: MU_NEWBIE / 6., // sigma/2
+            convergence_eps: 2e-4,
+            sigma_growth: 5.,
+            rating: TSRating::new(),
+        }
+    }
+}
+
+fn gen_team_message<T, K: Clone>(places: &Vec<Vec<T>>, default: &K) -> Vec<Vec<K>> {
+    places
+        .iter()
+        .map(|place| vec![default.clone(); place.len()])
+        .collect()
+}
+
+fn gen_player_message<T, K: Clone>(places: &Vec<Vec<Vec<T>>>, default: &K) -> Vec<Vec<Vec<K>>> {
+    let mut ret = Vec::with_capacity(places.len());
+
+    for place in places {
+        ret.push(Vec::with_capacity(place.len()));
+
+        for team in place {
+            ret.last_mut()
+                .unwrap()
+                .push(vec![default.clone(); team.len()]);
+        }
+    }
+
+    ret
+}
+
+fn infer1(who: &mut Vec<impl TreeNode>) {
+    for item in who {
+        item.infer();
+    }
+}
+
+fn infer2(who: &mut Vec<Vec<impl TreeNode>>) {
+    for item in who {
+        infer1(item);
+    }
+}
+
+fn infer3(who: &mut Vec<Vec<Vec<impl TreeNode>>>) {
+    for item in who {
+        infer2(item);
+    }
+}
+
+fn infer_ld(ld: &mut Vec<impl TreeNode>, l: &mut Vec<impl TreeNode>) {
+    for i in 0..ld.len() {
+        l[i].infer();
+        ld[i].infer();
+    }
+    l.last_mut().unwrap().infer();
+    for i in (0..ld.len()).rev() {
+        ld[i].infer();
+        l[i].infer();
+    }
+}
+
+fn check_convergence(
+    a: &[Rc<RefCell<(TSMessage, TSMessage)>>],
+    b: &[(TSMessage, TSMessage)],
+) -> f64 {
+    if a.len() != b.len() {
+        return INFINITY;
+    }
+
+    let mut ret = 0.;
+
+    for i in 0..a.len() {
+        ret = f64::max(
+            ret,
+            f64::max(
+                f64::max(
+                    f64::abs(RefCell::borrow(&a[i]).0.mu - b[i].0.mu),
+                    f64::abs(RefCell::borrow(&a[i]).0.sigma - b[i].0.sigma),
+                ),
+                f64::max(
+                    f64::abs(RefCell::borrow(&a[i]).1.mu - b[i].1.mu),
+                    f64::abs(RefCell::borrow(&a[i]).1.sigma - b[i].1.sigma),
+                ),
+            ),
+        );
+    }
+
+    ret
+}
+
+impl TrueSkillSPBSystem {
+    fn inference(&mut self, contest: &TSContest) {
+        if contest.is_empty() {
+            return;
+        }
+
+        // could be optimized, written that way for simplicity
+        let mut s = gen_player_message(contest, &ProdNode::new());
```

Comment:
Hey does `cargo clippy` complain about there being too many single-letter variable names? Let's try to make our code pass `clippy`.